### ISSUE TITLE
snapshots recorded during `cypress run` mismatch `cypress open`

### ISSUE THREAD

**Issue by goldylucks** [palmerhq/cypress-image-snapshot#95]
To reproduce:

1. write in a test `cy.matchImageSnapshot()`
1. run `$ node_modules/.bin/cypress run`
1. run again `$ node_modules/.bin/cypress run` - test passes
1. run `$ node_modules/.bin/cypress open`
1. click on the aforementioned test - test fails

This might be a bug/inconsistency on `cypress`'s end, so let me know if I should open it there as well.

Many thanks for this awesome tool!

**Comment by goldylucks** [comment]
Seems like `cypress run` only takes screenshot of the visible screen, and `cypress open` scrolls and screenshots the entire page

I recorded now using `cypress open` and then did `cypress run` and this is the error:

`Error: Image size (360x640) different than saved snapshot size (360x3566).`

**Comment by goldylucks** [comment]
hmmm ... recording with `cypress run` didn't wait for content to load, while `cypress open` did. Maybe that's why there's such a difference in image height.

![init diff](https://user-images.githubusercontent.com/6963212/62466914-2b8d3880-b79b-11e9-84ad-79abc0e6c9ba.png)

I will try to make cypress wait for content to load before taking the snapshot, and run the test again

**Comment by goldylucks** [comment]
I changed the capture to viewport:
`addMatchImageSnapshotCommand({ capture: 'viewport' })`

so both tests are same dimensions, but they are still failing, tho I don't see any difference in the diff image:

![init diff](https://user-images.githubusercontent.com/6963212/62510610-4e116700-b818-11e9-8c9d-351182f82c04.png)

In the following image, I think everything in the left picture is one pixel higher than the right image:

![picture-mode diff](https://user-images.githubusercontent.com/6963212/62510633-6d0ff900-b818-11e9-9f5f-dc4702df356a.png)

Any ideas why this is happening?

**Comment by goldylucks** [comment]
I ran it in gitlab ci, and there's a mismatch between the images taken by the ci and taken locally, with the same cypress command `cypress run`

**Comment by sebinsua** [comment]
You might have a different issue from me, but [I ran into an issue due to headed mode using a different device scale factor that what is configured for headless mode, and I was able to fix it like so.](https://github.com/cypress-io/cypress/issues/2102#issuecomment-521299946)

**Comment by goldylucks** [comment]
@sebinsua the failure percentage varies from 5-28(!) percent :/

your solution disables the image matching in a headed electron browser? This means it won't compare the snapshots when I do `cypress open` ?

**Comment by sebinsua** [comment]
> your solution disables the image matching in a headed electron browser? This means it won't compare the snapshots when I do `cypress open`?

Yes, but that is just because I can't control the device scale factor in headed mode, and without control of this it's meaningless to compare the images (they would always have a different size).

As I said, in your instance, I think you should increase the `failureThreshold`. Btw, I've also ran into similar issues with broken behaviours when screenshotting elements that leave the viewport.

**Comment by ivoiv** [comment]
I get the same issue.

cypress.json
```
{
  "viewportWidth": 1400,
  "viewportHeight": 660
}
````

commands.js
````
addMatchImageSnapshotCommand({
  failureThreshold: 0.003, // threshold for entire image
  failureThresholdType: "percent", // percent of image or number of pixels
  customDiffConfig: { threshold: 0.1 }, // threshold for each pixel
  capture: 'viewport', // capture viewport in screenshot
)}
````

When I run tests in cypress open I get a snapshot resolution of 1400x600.
When I run in cypress run I get a snapshot of 1280x660

One strange behaviour is that if i remove the viewport config from cypress.json i get a snapshot of 1000x660, which is the default viewport.

If i change cypress.json to
````
{
  "viewportWidth": 1390,
  "viewportHeight": 650
}
```

Then recorded snapshot in cypress run mode changes to 1280x650, so i reacts to viewportHeight but not to viewportWidth.

If you have a custom viewport setting, viewportWidth defaults to 1280.

**Comment by ivoiv** [comment]
Apparently this is a problem with Cypress and not with this plugin

https://github.com/cypress-io/cypress/issues/2102

**Comment by RichieRunner** [comment]
what can we do to get around this? what is the optimum viewport for both open and run to match?

**Comment by jsphstls** [comment]
My workaround is to only capture images in headless mode.
```
 Cypress.browser.isHeadless
    ? cy.matchImageSnapshot()
    : cy.log('Snapshot skipped in headed mode.')
```

**Comment by Karnikel** [comment]
Because I'm using the headed mode just for test developing...I used this workaround in my code:
```
    if (Cypress.browser.isHeadless) {
      cy.matchImageSnapshot('AnySnapshot');
    } else {
      cy.matchImageSnapshot('AnySnapshot', { failureThreshold: 0.1 });
      cy.log('Compared image with 10% failure threshold, because test is running in headed mode!');
      cy.log('PLEASE ADD NEW REFERENCE-SNAPSHOTS JUST IN HEADLESS MODE!');
    }
```

**Comment by jennifer-shehane** [comment]
I would refer to this issue in Cypress for this one: https://github.com/cypress-io/cypress/issues/3324

Cypress's current recommendation is to only take screenshots for comparison during `cypress run`. There's further explanation here: https://github.com/cypress-io/cypress/issues/3324#issuecomment-542414532

You could write logic to disable screenshotting during `cypress open` by using the `isInteractive` config option. 

```js
if (Cypress.config('isInteractive')) {
  // interactive "cypress open" mode
} else {
  // "cypress run" mode
}
```

**Comment by edouard-lopez** [comment]
Trying to overwrite `matchImageSnapshot`, I got the error:

  > Cannot overwite command for: matchImageSnapshot. An existing command does not exist by that name.

Overwriting `screenshot` I got the error:

	Image was NaN% different from saved snapshot with undefined different pixels.
	See diff for details: undefined

Asserted the headed vs headless mode seem to work but looks clumsy:

      Cypress.browser.isHeadless
        ? cy.matchImageSnapshot({
            customSnapshotIdentifier: 'in-iframe/conversation-is-open',
            failureThreshold: 0.03,
            failureThresholdType: 'percent', // percent of image or number of pixels
          })
        : cy.log('No screenshot taken when headed');

**Comment by harrync** [comment]
@edouard-lopez thanks for that, that's been the only way I've gotten it to be useable, what I did to clean it up a little was add that code to a custom Command, e.g.

In commands.js:
```js
Cypress.Commands.add('matchImageHeadless', () => {
  Cypress.browser.isHeadless
    ? cy.matchImageSnapshot({
        customSnapshotIdentifier: 'in-iframe/conversation-is-open',
        failureThreshold: 0.03,
        failureThresholdType: 'percent',
      })
    : cy.log('No screenshot taken when headed');
});
```

In *.spec.js:
```js
cy.matchImageHeadless();
```

Would love to be able to run the plugin in normal mode too though, but for now this works with our Husky setup

**Comment by MadnezGirl** [comment]
Hi, I have the same issue in my project.
I set headless mode instead headed like suggest Jennifer, also set chromium instead chrome.
Also I set config:
```
--force-device-scale-factor=1
--force-color-profile=srgb
--font-render-hinting=none
```

**But I still have issue with minimal difference in screenshots. Not usually in font but in images!**

My env:
```
Cypress 12.17.3
Node 18.12.0
Chromium 115 version
Running in docker with linux
```

![Uploading in-image-vertical-c-большим количеством товаров.diff.png…]()

Only changing the picture helped. I don’t know how long this will save the situation.
